File: src/main/java/cube8540/oauth/authentication/credentials/authority/application/DefaultSecuredResourceManagementService.java
package cube8540.oauth.authentication.credentials.authority.application;

import cube8540.oauth.authentication.credentials.authority.domain.AuthorityCode;
import cube8540.oauth.authentication.credentials.authority.domain.ResourceMethod;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResource;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResourceId;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResourceRepository;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResourceValidationPolicy;
import cube8540.oauth.authentication.credentials.authority.domain.exception.ResourceNotFoundException;
import cube8540.oauth.authentication.credentials.authority.domain.exception.ResourceRegisterException;
import lombok.Setter;

import java.net.URI;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class DefaultSecuredResourceManagementService implements SecuredResourceManagementService {

    private final SecuredResourceRepository repository;

    @Setter
    private SecuredResourceValidationPolicy policy;

    public DefaultSecuredResourceManagementService(SecuredResourceRepository repository) {
        this.repository = repository;
    }

    @Override
    public Long count(String resourceId) {
        return repository.countByResourceId(new SecuredResourceId(resourceId));
    }

    @Override
    public List<SecuredResourceDetails> getResources() {
        return repository.findAll().stream().map(DefaultSecuredResourceDetails::of).collect(Collectors.toList());
    }

    @Override
    public SecuredResourceDetails registerNewResource(SecuredResourceRegisterRequest registerRequest) {
        if (repository.countByResourceId(new SecuredResourceId(registerRequest.getResourceId())) > 0) {
            throw ResourceRegisterException.existsIdentifier(registerRequest.getResourceId() + " is already exists");
        }
        SecuredResource resource = new SecuredResource(new SecuredResourceId(registerRequest.getResourceId()),
                URI.create(registerRequest.getResource()), ResourceMethod.of(registerRequest.getMethod()));
        Optional.ofNullable(registerRequest.getAuthorities()).orElse(Collections.emptyList())
                .stream().map(AuthorityCode::new).forEach(resource::addAuthority);
        resource.validation(policy);
        return DefaultSecuredResourceDetails.of(repository.save(resource));
    }

    @Override
    public SecuredResourceDetails modifyResource(String resourceId, SecuredResourceModifyRequest modifyRequest) {
        SecuredResource resource = getResource(resourceId);
        resource.changeResourceInfo(URI.create(modifyRequest.getResource()), ResourceMethod.of(modifyRequest.getMethod()));
        Optional.ofNullable(modifyRequest.getRemoveAuthorities()).orElse(Collections.emptyList())
                .stream().map(AuthorityCode::new).forEach(resource::removeAuthority);
        Optional.ofNullable(modifyRequest.getNewAuthorities()).orElse(Collections.emptyList())
                .stream().map(AuthorityCode::new).forEach(resource::addAuthority);
        resource.validation(policy);
        return DefaultSecuredResourceDetails.of(repository.save(resource));
    }

    @Override
    public SecuredResourceDetails removeResource(String resourceId) {
        SecuredResource resource = getResource(resourceId);

        repository.delete(resource);
        return DefaultSecuredResourceDetails.of(resource);
    }

    private SecuredResource getResource(String resourceId) {
        return repository.findById(new SecuredResourceId(resourceId))
                .orElseThrow(() -> ResourceNotFoundException.instance(resourceId + " is not found"));
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/resource/application/DefaultSecuredResourceManagementService.kt
package cube8540.oauth.authentication.resource.application

import cube8540.oauth.authentication.resource.domain.*
import cube8540.oauth.authentication.resource.domain.ResourceNotFoundException.Companion.instance
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.net.URI

@Service
class DefaultSecuredResourceManagementService @Autowired constructor(private val repository: SecuredResourceRepository): SecuredResourceManagementService {

    @set:[Autowired Qualifier("defaultSecuredResourceValidatorFactory")]
    lateinit var validatorFactory: SecuredResourceValidatorFactory

    @Transactional
    override fun registerNewResource(registerRequest: SecuredResourceRegisterRequest): SecuredResourceDetails {
        if (count(registerRequest.resourceId) > 0) {
            throw ResourceRegisterException.existsIdentifier("${registerRequest.resourceId} is already exists")
        }

        val resource = SecuredResource(SecuredResourceId(registerRequest.resourceId),
            URI.create(registerRequest.resource), ResourceMethod.of(registerRequest.method))
        registerRequest.authorities?.forEach { resource.addAuthority(it.authority) }
        resource.validation(validatorFactory)
        return DefaultSecuredResourceDetails.of(repository.save(resource))
    }

    @Transactional
    override fun modifyResource(resourceId: String, modifyRequest: SecuredResourceModifyRequest): SecuredResourceDetails {
        val resource = getResource(resourceId)

        resource.changeResourceInfo(URI.create(modifyRequest.resource), ResourceMethod.of(modifyRequest.method))
        modifyRequest.removeAuthorities?.forEach { resource.removeAuthority(it.authority) }
        modifyRequest.newAuthorities?.forEach { resource.addAuthority(it.authority) }
        resource.validation(validatorFactory)
        return DefaultSecuredResourceDetails.of(repository.save(resource))
    }

    @Transactional
    override fun removeResource(resourceId: String): SecuredResourceDetails {
        val resource = getResource(resourceId)

        repository.delete(resource)
        return DefaultSecuredResourceDetails.of(resource)
    }

    @Transactional(readOnly = true)
    override fun count(resourceId: String): Long = repository.countByResourceId(SecuredResourceId(resourceId))

    @Transactional(readOnly = true)
    override fun getResources(): List<SecuredResourceEntry> =
        repository.findAll().map(SecuredResourceEntry::of).toList()

    private fun getResource(resourceId: String): SecuredResource =
        repository.findById(SecuredResourceId(resourceId))
            .orElseThrow { instance("$resourceId is not found") }
}